### PR TITLE
[Feature] Implementing the Deletion of Account.

### PR DESCRIPTION
**Pull request by anmode**
# Descripción

https://user-images.githubusercontent.com/88332977/227430571-1ef85566-f64f-4d7d-a523-0cd90042b2d6.mov

Fixes: #6945

# Comentarios
Done

# Checklist:

- [x] El código sigue la [guía de
      estilo](https://github.com/omegaup/omegaup/wiki/Coding-guidelines) de
      omegaUp.
- [x] Se corrieron todas las pruebas y pasaron.
- [x] Si se está agregando funcionalidad nueva, se agregaron pruebas.
- [x] Si el cambio es grande (> 200 líneas), hay que intentar partirlo en
      varios pull requests. De preferencia uno para los controladores + phpunit
      y luego otro para la interfaz.

### CODE REVIEW

File: frontend/www/js/omegaup/components/user/SidebarMainInfo.vue [hunk 1]
```diff
@@ -82,6 +100,7 @@ import country_Flag from '../CountryFlag.vue';
 import user_Username from './Username.vue';
 import { types } from '../../api_types';
 import { Problem } from '../../linkable_resource';
+// import { deleteRequest } from '../../api_types';
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/user/SidebarMainInfo.vue:103]: If this line is not used, it's better removing it

File: frontend/www/js/omegaup/components/user/SidebarMainInfo.vue [hunk 2]
```diff
@@ -71,7 +71,25 @@
...
       </a>
+      <div class="card-body text-center">
+        <!-- Add a button to delete the account -->
+        <button
+          class="btn btn-danger btn-sm my-1 w-100"
+          @click="showConfirmationModal = true"
+        >
+          {{ T.accountDelete }}
+        </button>
+      </div>
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/user/SidebarMainInfo.vue:82]: In order to add a new item in the menu, you don't need create the div and button, both of them are created automatically. 
```suggestion
```
So, you only need to add:
```ts
  { key: 'delete-account', title: T.userEditDeleteAccount, visible: true },
```
in `urlMapping` object

**Comment by tvanessa** [comment]
As @pabo99 [suggested](https://github.com/omegaup/omegaup/issues/6945#issuecomment-1475711874) I also prefer reusing the existing component for deleting problems.

File: frontend/www/js/omegaup/components/user/SidebarMainInfo.vue [hunk 3]
```diff
@@ -112,6 +131,7 @@ export default class UserSidebarMainInfo extends Vue {
     this.selectedTab,
     this.currentUrlMapping,
   );
+  showConfirmationModal = false;
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/user/SidebarMainInfo.vue:134]:
```suggestion
```

**Review by pabo99 — COMMENTED** [review]
I have some suggestions for you on how you could implement the feature. If you have any doubts, don't hesitate to ask us.

File: frontend/www/js/omegaup/components/user/SidebarMainInfo.vue [hunk 4]
```diff
@@ -71,7 +71,25 @@
...
     </div>
+    <b-modal
+      v-model="showConfirmationModal"
+      title="Confirmation required"
+      ok-title="Delete"
+      cancel-title="Cancel"
+      @ok="deleteAccount"
+    >
+      <p>Are you sure you want to delete your account?</p>
+    </b-modal>
```
[the remark this answers is not in the document]
Comment (heduenas) [reply frontend/www/js/omegaup/components/user/SidebarMainInfo.vue:92]: I agree with @pabo99's suggestion for the placement of the button. However, I think we should keep the modal to ask for user confirmation.

[the remark this answers is not in the document]
Comment (heduenas) [reply frontend/www/js/omegaup/components/user/SidebarMainInfo.vue:92]: I have also filed https://github.com/omegaup/omegaup/issues/6949 to do the same thing for problem deletion.

File: frontend/www/js/omegaup/components/user/SidebarMainInfo.vue (hunk 2 above)
Comment (anmode) [reply frontend/www/js/omegaup/components/user/SidebarMainInfo.vue:82]: ohh Okayy.

File: frontend/www/js/omegaup/components/user/SidebarMainInfo.vue (hunk 1 above)
Comment (anmode) [reply frontend/www/js/omegaup/components/user/SidebarMainInfo.vue:103]: sure!!

**Comment by anmode** [comment]
Thanks @pabo99 , @heduenas / @tvanessa Could help me with the description part? so I will make changes in the translation strings.

File: frontend/www/js/omegaup/components/user/DeleteAccount.vue [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+          <omegaup-markdown
+            :markdown="T.wordsDangerZoneDesc"
+          ></omegaup-markdown>
+          <br /><br />
+          <button
+            class="btn btn-danger"
+            type="submit"
+            @click="showConfirmationModal = true"
+          >
+            {{ T.wordsDelete }}
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/user/DeleteAccount.vue:17]: Here you need to change the text in th button for the following string:

EN:
```
Request user account deletion
```
ES:
```
Solicitar eliminación de cuenta de usuario
```
PT:
```
Solicitar exclusão de conta de usuário
```

File: frontend/www/js/omegaup/components/user/DeleteAccount.vue [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+          <h4 class="alert-heading">{{ T.wordsDangerZone }}</h4>
+          <hr />
+          <omegaup-markdown
+            :markdown="T.wordsDangerZoneDesc"
+          ></omegaup-markdown>
+          <br /><br />
+          <button
+            class="btn btn-danger"
+            type="submit"
+            @click="showConfirmationModal = true"
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/user/DeleteAccount.vue:15]: In addition to displaying the modal dialog, you need to make an API call to `User::apiDeleteRequest` in order to obtain the token required to delete the user account.

File: frontend/www/js/omegaup/components/user/DeleteAccount.vue [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+          >
+            {{ T.wordsDelete }}
+          </button>
+        </div>
+      </div>
+    </form>
+    <b-modal
+      v-model="showConfirmationModal"
+      title="Confirmation required"
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/user/DeleteAccount.vue:24]: All the strings should be added in the lozalizated files

File: frontend/www/js/omegaup/components/user/Profile.vue [hunk 8]
```diff
@@ -74,6 +74,12 @@
             @add-password="(request) => $emit('add-password', request)"
           ></omegaup-user-add-password>
         </template>
+        <template v-else-if="currentSelectedTab === 'delete-account'">
+          <omegaup-user-delete-account
+            :username="profile.username"
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/user/Profile.vue:79]: I'm not sure if we need the username, because we can call the APIs `User::apiDeleteRequest` and `User::apiDeleteConfirm` without this field. So you could remove this line

File: frontend/www/js/omegaup/components/user/Profile.vue [hunk 9]
```diff
@@ -74,6 +74,12 @@
...
+            @add-password="(request) => $emit('delete-account', request)"
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/user/Profile.vue:80]: nit:
```suggestion
            @request-delete-account="$emit('request-delete-account')"
```

File: frontend/www/js/omegaup/components/user/DeleteAccount.vue (hunk 6 above)
Comment (anmode) [reply frontend/www/js/omegaup/components/user/DeleteAccount.vue:15]: So the thing is when user click `Request user account deletion`. we will call `User::apiDeleteRequest`. and when user click the confirmation dailog box. it will call `User::apiDeleteConfirm`. isn't?

File: frontend/www/js/omegaup/components/user/Profile.vue (hunk 8 above)
Comment (anmode) [reply frontend/www/js/omegaup/components/user/Profile.vue:79]: But `User::apiDeleteRequest` and `User::apiDeleteConfirm`. requires username ig!!

File: frontend/www/js/omegaup/components/user/DeleteAccount.vue (hunk 7 above)
Comment (anmode) [reply frontend/www/js/omegaup/components/user/DeleteAccount.vue:24]: okayy

File: frontend/www/js/omegaup/components/user/DeleteAccount.vue (hunk 5 above)
Comment (anmode) [reply frontend/www/js/omegaup/components/user/DeleteAccount.vue:17]: ohh ok ok

File: frontend/www/js/omegaup/components/user/Profile.vue (hunk 9 above)
Comment (anmode) [reply frontend/www/js/omegaup/components/user/Profile.vue:80]: okk

File: frontend/www/js/omegaup/components/user/DeleteAccount.vue (hunk 6 above)
Comment (pabo99) [reply frontend/www/js/omegaup/components/user/DeleteAccount.vue:15]: That's right!

File: frontend/www/js/omegaup/components/user/Profile.vue (hunk 8 above)
Comment (pabo99) [reply frontend/www/js/omegaup/components/user/Profile.vue:79]: While it may seem like `username` is a required field for deleting a user account, if you take a look at the code

https://github.com/omegaup/omegaup/blob/7a1c46a9a041a25acbaf42ba67d8389670408929/frontend/server/src/Controllers/User.php#L3307-L3325

you'll see that it's actually optional. This is because only sysadmins need to provide a username to delete an account. On the other hand, when a regular user wants to delete their own account, the functions `resolveTargetIdentity` and `resolveTargetUser` retrieve the necessary information from the logged-in user."

Comment (anmode) [reply frontend/www/js/omegaup/components/user/Profile.vue:79]: ohh okayyy JP! got that

**Comment by anmode** [comment]
I updated the video in the description...Now only the Account delete Description is left. If i can get description I'll update it too

**Review by heduenas — COMMENTED** [review]
Can you file a new issue to sign off the user right after they confirm they want to delete their account? We should do that, but it's probably better to do it in a separate PR.

File: frontend/www/js/omegaup/components/user/DeleteAccount.vue [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+import T from '../../lang';
+import omegaup_Markdown from '../Markdown.vue';
+
+@Component({ components: { 'omegaup-markdown': omegaup_Markdown } })
+export default class UserDeleteAccount extends Vue {
+  T = T;
+  showConfirmationModal = false;
+  username = '';
+
+  // onConfirmDelete(): void {
```
Comment (heduenas) [inline frontend/www/js/omegaup/components/user/DeleteAccount.vue:46]: Delete this commented code?

File: frontend/www/js/omegaup/user/profile.ts [hunk 11]
```diff
@@ -189,6 +189,18 @@ OmegaUp.on('ready', () => {
               })
               .catch(ui.apiError);
           },
+          'request-delete-account': () => {
+            api.User.deleteRequest()
+              .then(({ token }) => {
+                api.User.deleteConfirm({ token })
+                  .then(() => {
+                    ui.success(T.requestAccountDeleted);
+                    location.reload(); // Reload the window
```
Comment (heduenas) [inline frontend/www/js/omegaup/user/profile.ts:198]: Right before reloading the window we want to log out the user. You can either do that here or in a separate PR.

**Comment by anmode** [comment]
> Can you file a new issue to sign off the user right after they confirm they want to delete their account? We should do that, but it's probably better to do it in a separate PR.

sure!

File: frontend/www/js/omegaup/components/user/DeleteAccount.vue (hunk 10 above)
Comment (anmode) [reply frontend/www/js/omegaup/components/user/DeleteAccount.vue:46]: oops! my bad. I forgot about it. ok removing it

File: frontend/www/js/omegaup/user/profile.ts (hunk 11 above)
Comment (anmode) [reply frontend/www/js/omegaup/user/profile.ts:198]: sure! I'll continue in pR for sign off the user right

**Comment by anmode** [comment]
One thing is left that description part!! Please can you give my description in English and i will update the translation string accordingly

**Comment by pabo99** [comment]
@tvanessa could you please help us with the descriptions displayed on this view?

**Comment by anmode** [comment]
https://user-images.githubusercontent.com/88332977/227777485-e8dc7fc6-d4f8-457e-8fdb-4bc8ae22e66d.mov

Done! @heduenas

**Comment by heduenas** [comment]
@anmode can you please fix the linter?

File: frontend/templates/es.lang [hunk 12]
```diff
@@ -1,6 +1,10 @@
 U13CannotPerform = "Usuario menor de 13 años no puede realizar esta acción."
 aboutToStart = "Estás a punto de iniciar el concurso. Una vez que hagas clic en entrar, tu tiempo iniciará."
 acceptPrivacyPolicy = "He leído y acepto la [Política de Uso y Privacidad](http://blog.omegaup.com/privacy-policy/) del sitio, así como el [Código de Conducta](https://blog.omegaup.com/codigo-de-conducta-en-omegaup/)."
+accountDeleteCancel = "Cancelar"
+accountDeleteConfirmationMessage = "¿Estás segura de que quieres eliminar tu cuenta?"
+accountDeleteOk = "Eliminar"
+accountDeleteRequireConfirmation = "confirmación requerida"
```
Comment (heduenas) [inline frontend/templates/es.lang:7]: Capital Case this string.

File: frontend/templates/pt.lang [hunk 13]
```diff
@@ -1,6 +1,10 @@
 U13CannotPerform = "Usuário menor de 13 anos não pode realizar esta ação."
 aboutToStart = "Você está prestes a iniciar o concurso. Seu tempo começará depois de clicar em Começar Concurso."
 acceptPrivacyPolicy = "Li e aceito a [Política de Uso e Privacidade](http://blog.omegaup.com/privacy-policy/) do site, bem como o [Código de Conduta](https://blog.omegaup.com/codigo-de-conducta-en-omegaup/)."
+accountDeleteCancel = "Cancelar"
+accountDeleteConfirmationMessage = "Tem certeza de que deseja excluir sua conta?"
+accountDeleteOk = "Excluir"
+accountDeleteRequireConfirmation = "Necessária confirmação"
```
Comment (heduenas) [inline frontend/templates/pt.lang:7]: Capital Case.

File: frontend/templates/en.lang [hunk 14]
```diff
@@ -1252,6 +1256,7 @@ reportProblemFormPoorlyDescribed = "The problem is poorly described."
 reportProblemFormQuestion = "Why do you want to report this problem?"
 reportProblemFormSpam = "Is Spam."
 reportProblemFormThanksForReview = "Thanks for your report!"
+requestAccountDeleted = "Your Account has been successfully deleted"
```
Comment (heduenas) [inline frontend/templates/en.lang:1259]: "Your account has been successfully deleted"

File: frontend/templates/es.lang [hunk 15]
```diff
@@ -1400,6 +1405,7 @@ userEditChangePasswordOldPassword = "Contraseña anterior"
 userEditChangePasswordRepeatNewPassword = "Repite la contraseña de nuevo"
 userEditChangePasswordToUsername = "Cambiar contraseña: %(username)"
 userEditCountry = "País / Región"
+userEditDeleteAccount = "Eliminar cuenta"
```
Comment (heduenas) [inline frontend/templates/es.lang:1408]: Capital Case

File: frontend/templates/pt.lang [hunk 16]
```diff
@@ -1400,6 +1405,7 @@ userEditChangePasswordOldPassword = "Senha antiga"
 userEditChangePasswordRepeatNewPassword = "Repita nova senha"
 userEditChangePasswordToUsername = "Alterar senha: %(username)"
 userEditCountry = "País / Região"
+userEditDeleteAccount = "Deletar conta"
```
Comment (heduenas) [inline frontend/templates/pt.lang:1408]: Capital Case.

File: frontend/www/js/omegaup/user/profile.ts [hunk 17]
```diff
@@ -189,6 +189,20 @@ OmegaUp.on('ready', () => {
               })
               .catch(ui.apiError);
           },
+          'request-delete-account': () => {
+            api.User.deleteRequest()
+              .then(({ token }) => {
+                api.User.deleteConfirm({ token })
+                  .then(() => {
+                       // Reload the window
```
Comment (heduenas) [inline frontend/www/js/omegaup/user/profile.ts:197]: Incorrect indentation?

File: frontend/www/js/omegaup/components/user/Profile.vue (hunk 8 above)
Comment (heduenas) [reply frontend/www/js/omegaup/components/user/Profile.vue:79]: Let us know when you remove the username parameter here to review this part again.

File: frontend/www/js/omegaup/user/profile.ts [hunk 18]
```diff
@@ -194,8 +194,10 @@ OmegaUp.on('ready', () => {
               .then(({ token }) => {
                 api.User.deleteConfirm({ token })
                   .then(() => {
-                    ui.success(T.requestAccountDeleted);
-                    location.reload(); // Reload the window
+                       // Reload the window
+                       location.reload();
```
Comment (pabo99) [inline frontend/www/js/omegaup/user/profile.ts:198]:
```suggestion
```

File: frontend/templates/pt.lang (hunk 16 above)
Comment (pabo99) [reply frontend/templates/pt.lang:1408]: "Delete Account", right @heduenas ?

Comment (anmode) [reply frontend/templates/pt.lang:1408]: Ohh okayy got it! I misunderstood Lol... thanks

Comment (heduenas) [reply frontend/templates/pt.lang:1408]: I meant Title Case, sorry.

**Comment by heduenas** [comment]
@anmode here's the description text for the delete account view: 

EN:
Click the following button if you wish to delete your account and all your personal information from omegaUp. This operation will take place immediately after you confirm that you want to do it and it is not reversible.

ES:
Haz clic en el siguiente botón si deseas borrar tu cuenta y toda tu información personal de omegaUp. Esta operación se realizará de forma inmediata una vez que confirmes que lo quieres hacer y no es reversible.

PT:
Clique no botão abaixo se desejar excluir sua conta e todas as suas informações pessoais do omegaUp. Esta operação será realizada imediatamente assim que você confirmar que deseja realizá-la e não é reversível.

**Comment by anmode** [comment]
@pabo99 Can we do something to increase the speed of lint?? If you can provide some resources. I would love to work on this. as really it took lot time on my machine

**Comment by anmode** [comment]
> @anmode here's the description text for the delete account view:
> 
> EN: Click the following button if you wish to delete your account and all your personal information from omegaUp. This operation will take place immediately after you confirm that you want to do it and it is not reversible.
> 
> ES: Haz clic en el siguiente botón si deseas borrar tu cuenta y toda tu información personal de omegaUp. Esta operación se realizará de forma inmediata una vez que confirmes que lo quieres hacer y no es reversible.
> 
> PT: Clique no botão abaixo se desejar excluir sua conta e todas as suas informações pessoais do omegaUp. Esta operação será realizada imediatamente assim que você confirmar que deseja realizá-la e não é reversível.

Thank you so much! I did the changes!!

**Comment by pabo99** [comment]
> @pabo99 Can we do something to increase the speed of lint?? If you can provide some resources. I would love to work on this. as really it took lot time on my machine

I'm not sure why the linter process is so slow in some environments. While it works fine on my computer, I'm unable to determine the cause of the slowness and how to speed it up

**Review by pabo99 — APPROVED** [review]
LGTM

File: frontend/templates/es.lang [hunk 19]
```diff
@@ -1,6 +1,11 @@
 U13CannotPerform = "Usuario menor de 13 años no puede realizar esta acción."
 aboutToStart = "Estás a punto de iniciar el concurso. Una vez que hagas clic en entrar, tu tiempo iniciará."
 acceptPrivacyPolicy = "He leído y acepto la [Política de Uso y Privacidad](http://blog.omegaup.com/privacy-policy/) del sitio, así como el [Código de Conducta](https://blog.omegaup.com/codigo-de-conducta-en-omegaup/)."
+accountDeleteCancel = "Cancelar"
+accountDeleteConfirmationMessage = "¿Estás segura de que quieres eliminar tu cuenta?"
+accountDeleteMessage = "Click the following button if you wish to delete your account and all your personal information from omegaUp. This operation will take place immediately after you confirm that you want to do it and it is not reversible."
```
Comment (heduenas) [inline frontend/templates/es.lang:6]: Paste the spanish version in here.

File: frontend/templates/es.lang (hunk 12 above)
Comment (heduenas) [reply frontend/templates/es.lang:8]: Title Case rather

File: frontend/templates/pt.lang (hunk 13 above)
Comment (heduenas) [reply frontend/templates/pt.lang:8]: Title Case rather

**Review by heduenas — APPROVED** [review]
